lib/Conversion/TritonGPUToLLVM/AssertOpToLLVM.cpp

```diff
@@ -35,6 +35,11 @@ struct AssertOpConversion : public ConvertOpToLLVMPattern<triton::AssertOp> {
       }
     }
     llAssert(op, condition, adaptor.getMessage(), rewriter);
+    // Add a barrier to avoid a race condition in case an assert is followed by
+    // an op that may trap if the assert condition is true. Since the tensor in
+    // those two operations may have different layout we need to make sure all
+    // the threads are done executing the assert before going to the next op.
+    barrier();
```

Comment:
Can we only insert the barrier for `RankedTensor` conditionals. That still fixes the issue, avoids an unnecessary barrier for scalar conds and fixes the hang with reductions all in one go.

Reply:
ah yeah sounds like an easy enough solution

Reply:
https://github.com/triton-lang/triton/pull/5043